src/client/websocket/handlers/TYPING_START.js

```diff
@@ -5,14 +5,62 @@ const { Events } = require('../../../util/Constants');
 module.exports = (client, { d: data }) => {
   const channel = client.channels.cache.get(data.channel_id);
   const user = client.users.cache.get(data.user_id);
+  const timestamp = new Date(data.timestamp * 1000);
 
   if (channel && user) {
-    /**
-     * Emitted whenever a user starts typing in a channel.
-     * @event Client#typingStart
-     * @param {Channel} channel The channel the user started typing in
-     * @param {User} user The user that started typing
-     */
-    client.emit(Events.TYPING_START, channel, user);
+    if (channel.type === 'voice') {
+      client.emit(Events.WARN, `Discord sent a typing packet to a voice channel ${channel.id}`);
+      return;
+    }
+
+    if (channel._typing.has(user.id)) {
+      const typing = channel._typing.get(user.id);
+
+      typing.lastTimestamp = timestamp;
+      typing.resetTimeout(tooLate(channel, user));
+    } else {
+      channel._typing.set(user.id, new TypingData(client, timestamp, tooLate(channel, user)));
+
+      /**
+       * Emitted whenever a user starts typing in a channel.
+       * @event Client#typingStart
+       * @param {Channel} channel The channel the user started typing in
+       * @param {User} user The user that started typing
+       */
+      client.emit(Events.TYPING_START, channel, user, channel._typing.get(user.id));
+    }
   }
 };
+
+/**
+ * Instance which keeps track of how long a user has been typing for.
+ */
+class TypingData {
```

Comment:
There is very little being done in this class here, should think about removing it and just bring the functionality up to the relevant places

Reply:
I've updated it so that it doesn't need the `TypingData` instance anymore. Let me know if there's anything else that needs to be done/updated. :+1:

Reply:
I've actually noticed a bit of redundancy which is an unnecessary function called `getElapsedTime`, I could easily just calculate the time without using it, do you want me to remove it?

Reply:
I've removed the `getElapsedTime` function for good, believe that it was kind of useless as the time was easily calculable through `channel._typing.get(...).since` and `Date.now()`.